### PR TITLE
fix(v2): removes hardcoded sidebar width

### PR DESCRIPTION
**Pull request by tusharf5**
## Motivation

This PR removes the hardcoded `width:20rem` value in the sidebar css in D2. 

Also, it was brought up in [#1530](https://github.com/facebook/Docusaurus/issues/1530)

### Have you read the [Contributing Guidelines on pull requests](https://github.com/facebook/Docusaurus/blob/master/CONTRIBUTING.md#pull-requests)?

Yes

## Test Plan

Local Testing.

- #### Before removing sidebar width
Before Removing Width

- #### After removing sidebar width
After Removing Width

## Related PRs

None

### CODE REVIEW

**Comment by tusharf5** [comment]
Also, I noticed that removing the hardcoded width leaves a lot of whitespace.

With the html structure as (**3 by 9 ratio**)

```html
<div>
  <div class="col col--3">SideBar<div>
  <div  class="col col--9">Main Content</div>
</div>
```

The page looks like this right now

Extra White Space

If we modify the classes to have (**2 by 10 ratio**)

```html
<div>
  <div class="col col--2">SideBar<div>
  <div  class="col col--10">Main Content</div>
</div>
```

Then the page looks like below

Normal

**Comment by wgao19** [comment]
#1530

**Comment by tusharf5** [comment]
Just noticed that this will cause the menu item text to overflow when it is too long as it won't inherit any width from the parent (**.sidebar**) as we will have removed it. That means we need to set some width on the **.sidebar** class. [See Sidebar in Image Below]

Overflow

So I'd suggest that instead of removing `width: 20rem;` we set its (**.sidebar**) width as `width: var(--ifm-col-width);` where `--ifm-col-width` is the same variable used to set width of its(**.sidebar**) parent i.e `col col--3`. This will resolve this overflowing issue too and also the width won't be harcoded. 

Another suggestion would be to use 2(sidebar) by 10(main) ratio rather than 3(sidebar) by 9(main). I've added pictures of that in my last comment.

**Comment by yangshun** [comment]
Thanks for the PR! While I appreciate the initiative in helping out, there are good reasons why we didn't add the "help wanted" label to the issue, e.g for the above reasons you mentioned and it could involve fixing the CSS library that we are using internally (which is not open sourced yet) instead of Docusaurus code.

I've been working on a fix and agree that this issue is not easy to fix. I haven't had time to come up with a proper solution yet, and since you're keen on working on it and have spent significant time looking into it, I'll let you take over. Do be prepared that this PR will take a while to merge with a few rounds of back-and-forth as this issue is pretty tricky. I'll follow up with a suggested approach later on.

In future, we would appreciate that you drop us a note before jumping on issues so that efforts don't overlap or end up non-fruitful. Thank you for the understanding!

**Comment by tusharf5** [comment]
> I've been working on a fix and agree that this issue is not easy to fix. I haven't had time to come up with a proper solution yet, and since you're keen on working on it and have spent significant time looking into it, I'll let you take over. Do be prepared that this PR will take a while to merge with a few rounds of back-and-forth as this issue is pretty tricky. I'll follow up with a suggested approach later on.

My bad, I should have first asked on the issue thread before spending time working on it. I am still new to open source so everything is pretty much a learning for me :)  You can comment on this thread and I will follow up with you.

> In future, we would appreciate that you drop us a note before jumping on issues so that efforts don't overlap or end up non-fruitful. 

I'll take note of that 👍

 Thanks for the valuable feedback @yangshun

**Comment by endiliey** [comment]
I personally like the 3-9 ratio.

My fix suggestion is to hide the nav for situation described in #1530

We will most likely need to work on the sidebar menu on mobile as well.

what do yall think

**Comment by wgao19** [comment]
Hi @tusharf5,

Thanks for helping and sorry about the delayed reply. 

The page's grid layout is intended to give the sidebar a 25% space. But sidebar is not respecting that due to its `position: fixed`, its (external) width calculation is gonna on viewport. And we do want to keep the `position: fixed` to keep the scrolling and positioning behavior. Therefore the fluid layout is not used.

So please keep the fixed width and remove the grid layout instead. Please also make sure that the page layout is stable on edge cases such as no content or very long texts on the sidebar, then we're good to land.

---
Just as a reference I'll also drop my notes here. Wanted to look at how other non-Docusaurus sites are doing it.

[RequireJS](https://requirejs.org/) uses a fixed width, `width: 220px`, but that is likely due to its design that needs a specific width

image

[w3.org](https://www.w3.org/TR/css-flexbox-1/) also uses a fixed with of `23em`. It does have a `max-width` but is computed to`calc(100% - 58px)`, probably for mobile.

image


[ReactJS](https://reactjs.org/docs/getting-started.html) is a `flex` on `300px`

Screen Shot 2019-06-04 at 10 36 43 AM

**Comment by tusharf5** [comment]
@wgao19 I've removed the flex layout and bring back fixed width sidebar. I had to make little changes in CSS at some places to accommodate this change. Please have a look.

File: packages/docusaurus-plugin-content-docs/src/theme/DocItem/styles.module.css [hunk 1]
```diff
@@ -9,6 +9,16 @@
   min-height: calc(100vh - 50px);
 }
 
+.container {
```
Comment (wgao19) [inline packages/docusaurus-plugin-content-docs/src/theme/DocItem/styles.module.css]: Can we name this more specifically to this component, like `.docItem`?

File: packages/docusaurus-plugin-content-docs/src/theme/DocItem/styles.module.css [hunk 2]
```diff
@@ -9,6 +9,16 @@
...
+  padding-top: 2rem !important;
```
Comment (wgao19) [inline packages/docusaurus-plugin-content-docs/src/theme/DocItem/styles.module.css]: Remove `! important` please.

File: packages/docusaurus-plugin-content-docs/src/theme/DocPage/styles.css [hunk 3]
```diff
@@ -0,0 +1,23 @@
+/**
+ * Copyright (c) 2017-present, Facebook, Inc.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+ /**
+ * width 19rem is calculated according to (sidebarWidth - sidebarPadding)
```
Comment (wgao19) [inline packages/docusaurus-plugin-content-docs/src/theme/DocPage/styles.css]: Thanks for the comment!

File: packages/docusaurus-plugin-content-docs/src/theme/DocPage/index.js [hunk 4]
```diff
@@ -12,6 +12,8 @@ import Layout from '@theme/Layout'; // eslint-disable-line
 
 import DocSidebar from '@theme/DocSidebar';
 
+import './styles.css';
```
Comment (wgao19) [inline packages/docusaurus-plugin-content-docs/src/theme/DocPage/index.js]: Let's be consistent and use CSS Modules here as well. Just leave out the class names that come from the dependency library.

File: packages/docusaurus-plugin-content-docs/src/theme/DocPage/styles.css [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+ * i.e 20rem - ( 0.5rem + 0.5rem )
+ * Todo - 0.5rem and 20rem should be a css variable
+ */
+
+ @media (min-width: 996px) {
+  .sidebar__container {
```
Comment (wgao19) [inline packages/docusaurus-plugin-content-docs/src/theme/DocPage/styles.css]: Move these two classes to CSS Modules please.

**Comment by wgao19** [comment]
Hey @tusharf5 thanks for the updates! A few tweaks and then we're good to go :)

**Comment by endiliey** [comment]
Thanks for the help 🔥

**Pull request merged**